File: server/service/src/main/kotlin/senegai/server/service/silvaoptionum/SilvaOptionumService.kt
/* @tt{{{

    @move-comment-backward
    @template-renderer [
        templateRendererClassName="ItemServiceRenderer"
        templateRendererPackageName="senegai.codegen.renderer.be"
        templateRendererInterfaceName="BeItemRenderer"
        templateRendererInterfacePackageName="senegai.codegen.renderer.be"
    ] [
        modelClassName="BeItemModel"
        modelPackageName="senegai.codegen.renderer.model.be"
        modelName="model"
    ]

    @replace-value-by-expression
        [ searchValue="SilvaOptionum" replaceByExpression="model.itemName.pascalCase" ]
        [ searchValue="silvaOptionum" replaceByExpression="model.itemName.camelCase" ]
        [ searchValue="silvaoptionum" replaceByExpression="model.itemName.lowerCase" ]

    @replace-value-by-expression
        [ searchValue="indexUnicus" replaceByExpression="model.primaryKeyAttribute.attributeName.camelCase" ]
        [ searchValue="UUID" replaceByExpression="model.primaryKeyAttribute.kotlinAttributeType" ]

    @modify-provided-filepath-by-replacements

}}}@ */
package senegai.server.service.silvaoptionum

import org.springframework.stereotype.Service
import senegai.server.service.bo.SilvaOptionumBO
import senegai.server.service.bo.SilvaOptionumByIdsCriteriaBO
import senegai.server.service.bo.SilvaOptionumSearchCriteriaBO
/* @tt{{{   @if [ conditionExpression="model.hasUuidPrimaryKey"]  }}}@ */
import java.util.UUID
/* @tt{{{   @end-if  }}}@ */

/**
 * Business service of the SilvaOptionum business context. It is called by the REST layer
 * with [SilvaOptionumBO] business objects (never WTOs) and delegates persistence to the
 * [SilvaOptionumRepository] port.
 *
 * Every operation works on the whole [SilvaOptionumBO] aggregate as a single unit.
 */
@Service
class SilvaOptionumService(
    private val silvaOptionumRepository: SilvaOptionumRepository,
) {

    fun getSilvaOptionumList(): List<SilvaOptionumBO> = silvaOptionumRepository.findAll()

    fun getSilvaOptionumById(indexUnicus: UUID): SilvaOptionumBO? =
        silvaOptionumRepository.findById(indexUnicus)

    fun getSilvaOptionumListByIds(criteria: SilvaOptionumByIdsCriteriaBO): List<SilvaOptionumBO> =
        silvaOptionumRepository.findByIds(criteria)

    fun searchSilvaOptionumList(searchCriteria: SilvaOptionumSearchCriteriaBO): List<SilvaOptionumBO> =
        silvaOptionumRepository.search(searchCriteria)

    fun createSilvaOptionum(silvaOptionum: SilvaOptionumBO): SilvaOptionumBO {
        /* @tt{{{   @if [ conditionExpression="model.hasGeneratedPrimaryKey" ]  }}}@ */
        val toCreate = silvaOptionum.copy(indexUnicus = silvaOptionumRepository.nextId())
        return silvaOptionumRepository.save(toCreate)
        /* @tt{{{
            @else
            @replace-value-by-expression
                [ searchValue="UUID(0L, 0L)" replaceByExpression="model.unresolvablePrimaryKeyValueExpression" ]
        }}}@ */
        require(silvaOptionum.indexUnicus != UUID(0L, 0L)) {
            "The primary key of SilvaOptionum is supplied by the caller and must not be empty."
        }
        return silvaOptionumRepository.save(silvaOptionum)
        /* @tt{{{   @end-replace-value-by-expression @end-if  }}}@ */
    }

    fun updateSilvaOptionum(silvaOptionum: SilvaOptionumBO): SilvaOptionumBO =
        silvaOptionumRepository.save(silvaOptionum)

    fun deleteSilvaOptionum(indexUnicus: UUID) = silvaOptionumRepository.deleteById(indexUnicus)
}

File: client/src/app/service/silva-optionum.service.ts
/* @tt{{{
    

    @move-comment-backward
    @template-renderer [
        templateRendererClassName="ItemServiceRenderer"
        templateRendererPackageName="senegai.codegen.renderer.angular"
        templateRendererInterfaceName="UiItemRenderer"
        templateRendererInterfacePackageName="senegai.codegen.renderer.angular"
    ] [
        modelClassName="UiItemModel"
        modelPackageName="senegai.codegen.renderer.model.ui"
        modelName="model"
    ]

    @replace-value-by-expression
        [ searchValue="SilvaOptionum" replaceByExpression="model.itemName.pascalCase" ]
        [ searchValue="silvaOptionum" replaceByExpression="model.itemName.camelCase" ]
        [ searchValue="silva-optionum" replaceByExpression="model.itemName.kebabCase" ]
        [ searchValue="SILVA_OPTIONUM" replaceByExpression="model.itemName.screamingSnakeCase" ]

    @replace-value-by-expression
        [ searchValue="indexUnicus" replaceByExpression="model.primaryKeyAttribute.attributeName.camelCase" ]
        [ searchValue="UUID" replaceByExpression="model.primaryKeyAttribute.typescriptAttributeType" ]


    @modify-provided-filepath-by-replacements

    

}}}@ */

import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {SilvaOptionumWTO} from "@app/wto/silva-optionum.wto";
import {SilvaOptionumSearchCriteriaWTO} from "@app/wto/silva-optionum-search-criteria.wto";
import {SilvaOptionumSearchResultWTO} from "@app/wto/silva-optionum-search-result.wto";
import {SilvaOptionumByIdsCriteriaWTO} from "@app/wto/silva-optionum-by-ids-criteria.wto";
import {SilvaOptionumByIdsResultWTO} from "@app/wto/silva-optionum-by-ids-result.wto";
/* @tt{{{   @if [ conditionExpression="model.hasUuidPrimaryKey"]  }}}@ */
import {UUID} from "@app/shared/uuid";
/* @tt{{{   @end-if  }}}@ */


@Injectable({providedIn: 'root'})
export class SilvaOptionumService {
    private readonly baseUrl = 'http://localhost:8081/api/silva-optionum';

    constructor(private readonly http: HttpClient) {}

    getSilvaOptionumList(): Observable<SilvaOptionumWTO[]> {
        return this.http.get<SilvaOptionumWTO[]>(this.baseUrl);
    }

    searchSilvaOptionumList(searchCriteria: SilvaOptionumSearchCriteriaWTO): Observable<SilvaOptionumSearchResultWTO> {
        return this.http.post<SilvaOptionumSearchResultWTO>(`${this.baseUrl}/search`, searchCriteria);
    }

    /**
     * Resolves a whole set of references to this item in one call, so that a list of stored
     * identifiers can be shown by the display attributes instead of the bare primary keys.
     */
    getSilvaOptionumListByIds(criteria: SilvaOptionumByIdsCriteriaWTO): Observable<SilvaOptionumByIdsResultWTO> {
        return this.http.post<SilvaOptionumByIdsResultWTO>(`${this.baseUrl}/by-ids`, criteria);
    }

    getSilvaOptionumById(indexUnicus: UUID): Observable<SilvaOptionumWTO | null> {
        return this.http.get<SilvaOptionumWTO | null>(`${this.baseUrl}/${indexUnicus}`);
    }

    deleteSilvaOptionum(indexUnicus: UUID): Observable<void> {
        return this.http.delete<void>(`${this.baseUrl}/${indexUnicus}`);
    }

    updateSilvaOptionum(silvaOptionum: SilvaOptionumWTO): Observable<SilvaOptionumWTO> {
        return this.http.put<SilvaOptionumWTO>(`${this.baseUrl}/${silvaOptionum.indexUnicus}`, silvaOptionum);
    }

    createSilvaOptionum(silvaOptionum: SilvaOptionumWTO): Observable<SilvaOptionumWTO> {
        return this.http.post<SilvaOptionumWTO>(this.baseUrl, silvaOptionum);
    }
}
